### PR TITLE
Cppcheck 2.15

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

Prompted by recent failures in our cppcheck workflow, use the latest version, and fix more issues that are being caught.

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: .github/workflows/cppcheck.yml [hunk 1]
```diff
@@ -55,11 +60,11 @@ jobs:
...
-            -i src/cli/test \
-            -i src/airflow/contam \
-            -i src/polypartition \
-            -i src/nano \
-            ./src \
+            -i src/cli/test/ \
+            -i src/airflow/contam/ \
+            -i src/polypartition/ \
+            -i src/nano/ \
+            . \
```
Comment (jmarrec) [inline .github/workflows/cppcheck.yml:67]: This effectively adds the root directories "python" and "ruby" to cppcheck

File: src/alfalfa/AlfalfaJSON.hpp [hunk 2]
```diff
@@ -100,7 +100,7 @@ namespace alfalfa {
     /**
      * Get a vector of all points currently exported to the Alfalfa API
      */
-    std::vector<AlfalfaPoint> points();
+    std::vector<AlfalfaPoint> points() const;
```
Comment (jmarrec) [inline src/alfalfa/AlfalfaJSON.hpp:103]: cppcheck 2.13 already complains about this one on develop right now

File: src/model/Building.hpp [hunk 3]
```diff
@@ -48,7 +48,7 @@ namespace model {
     /** @name Constructors and Destructors */
     //@{
 
-    virtual ~Building() = default;
+    virtual ~Building() override = default;
```
Comment (jmarrec) [inline src/model/Building.hpp:51]: lots of this.

File: src/model/SetpointManagerColdest.hpp [hunk 4]
```diff
@@ -48,22 +48,16 @@ namespace model {
     /** @name Getters */
     //@{
 
-    std::string controlVariable() const;
```
Comment (jmarrec) [inline src/model/SetpointManagerColdest.hpp:51]: Lots of `duplInheritedMembers` fixes

File: src/model/StraightComponent.hpp [hunk 5]
```diff
@@ -54,16 +52,10 @@ namespace model {
...
-
-    bool addToNode(Node& node);
-
-    ModelObject clone(Model model) const;
-
-    void disconnect();
+     *
+     *  Reimplemented from HVACComponent.
+     */
+    boost::optional<AirLoopHVAC> airLoopHVAC() const; // cppcheck-suppress [duplInheritedMember] for documentation purposes
```
Comment (jmarrec) [inline src/model/StraightComponent.hpp:58]: Some I keep, for documentation purposes mostly, because the docstring is helpful versus the base class one.

File: src/install_utility/main.cpp [hunk 6]
```diff
@@ -38,13 +38,12 @@ inline std::string applicationFilePath() {
   return {};
 }
 
-int main(int argc, char* argv[]) {
+int main(int argc, const char* argv[]) {
```
Comment (jmarrec) [inline src/install_utility/main.cpp:41]: I think this entire folder is leftover of some kind and should be removed

File: src/energyplus/Test/Space_GTest.cpp [hunk 7]
```diff
@@ -50,7 +50,7 @@
 
 #include "../../utilities/idf/IdfExtensibleGroup.hpp"
 #include "../../utilities/idf/WorkspaceExtensibleGroup.hpp"
-#include "utilities/core/Compare.hpp"
+#include "../../utilities/core/Compare.hpp"
```
Comment (jmarrec) [inline src/energyplus/Test/Space_GTest.cpp:53]: another example

File: src/utilities/geometry/FloorplanJS.hpp [hunk 8]
```diff
@@ -191,9 +191,9 @@ class UTILITIES_API FloorplanJS
   // recursively traverses through value and finds the largest id
   void setLastId(const Json::Value& value);
 
-  Json::Value* findByHandleString(Json::Value& value, const std::string& key, const std::string& handleString);
-  Json::Value* findByName(Json::Value& value, const std::string& key, const std::string& name, bool requireEmptyHandle);
-  Json::Value* findById(Json::Value& value, const std::string& key, const std::string& id);
+  static Json::Value* findByHandleString(Json::Value& value, const std::string& key, const std::string& handleString);
+  static Json::Value* findByName(Json::Value& value, const std::string& key, const std::string& name, bool requireEmptyHandle);
+  static Json::Value* findById(Json::Value& value, const std::string& key, const std::string& id);
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/geometry/FloorplanJS.hpp:196]: You might think this may be a terrible idea. Because Ruby bindings.

SWIG will no longer let you do `fs = FloorPlanJS(); fs.findByName(...)`

But that's in the "private" section, so all good

File: src/utilities/geometry/Polyhedron.cpp [hunk 9]
```diff
@@ -396,7 +396,7 @@ std::vector<Surface3dEdge> Polyhedron::edgesNotTwo(bool includeCreatedEdges) con
       }
     }
   }
-  for (auto& edge : edgesNotTwo) {
+  for (const auto& edge : edgesNotTwo) {
```
Comment (jmarrec) [inline src/utilities/geometry/Polyhedron.cpp:399]: const correctness

File: src/utilities/idd/IddFieldProperties.cpp [hunk 10]
```diff
@@ -184,7 +184,7 @@ std::ostream& IddFieldProperties::print(std::ostream& os) const {
       os << "       \\reference " << reference << '\n';
     }
   }
-  if (!references.empty()) {
+  if (!referenceClassNames.empty()) {
```
Comment (jmarrec) [inline src/utilities/idd/IddFieldProperties.cpp:187]: I think this was a mistake to begin with (copy paste)

File: src/utilities/idd/IddObject.cpp [hunk 11]
```diff
@@ -517,10 +517,9 @@ namespace detail {
   void IddObject_Impl::parseObject(const std::string& text) {
     // find the object name and the property text
     smatch matches;
-    string objectName;
     string propertiesText;
     if (boost::regex_search(text, matches, iddRegex::line())) {
-      objectName = string(matches[1].first, matches[1].second);
+      std::string objectName = string(matches[1].first, matches[1].second);
```
Comment (jmarrec) [inline src/utilities/idd/IddObject.cpp:522]: reducing scope

File: src/model/SubSurface.hpp [hunk 12]
```diff
@@ -176,6 +176,7 @@ namespace model {
     void resetAdjacentSubSurface();
 
     /** Returns the SurfacePropertyConvectionCoefficients, if it exists. */
+    // cppcheck-suppress [duplInheritedMember] because PlanarSurface is dumb and returns a vector
     boost::optional<SurfacePropertyConvectionCoefficients> surfacePropertyConvectionCoefficients() const;
```
Comment (jmarrec) [inline src/model/SubSurface.hpp:180]: I didn't want to break API, but this is clearly wrong.

This is wrong: https://github.com/NREL/OpenStudio/blob/62f977e804d9b6f6580dbedf760de621e2abf2ec/src/model/PlanarSurface.hpp#L199-L200

File: ruby/interpreter/RubyEval.hpp [hunk 13]
```diff
@@ -6,8 +6,8 @@
 #ifndef RUBYEVAL_included
 #define RUBYEVAL_included
 
-#include "ruby.h"
-#include "./RubyException.hpp"
+#include <ruby.h>
+#include "RubyException.hpp"
```
Comment (wenyikuang) [inline ruby/interpreter/RubyEval.hpp:10]: Just curios , why are using relative path above but keeps the hpp name here?

Comment (jmarrec) [reply ruby/interpreter/RubyEval.hpp:10]: I'm not sure I understand the question, could you rephrase please?
